### ISSUE TITLE
Cannot ping virtual ip address in RedHat Enterprise  Linux Server Release 7.4(Maipo)?

### ISSUE THREAD

**Issue by yarcl** [acassen/keepalived#1142]
It was puzzle me for a day, the virtual ip can be failover to the backup router, but can't be pinged by another machine.  It seems like the virtual ip can't be accessed!
Is there anyone meet this problem before?  Thanks for a lot!

**Comment by pqarmitage** [comment]
It rather depends on your configuration, which unfortunately you haven't posted.

The VRRPv3 RFC (RFC 5798) states in section 6.4.3 (regarding a master):
 ```
      (650) - MUST accept packets addressed to the IPvX address(es)
      associated with the virtual router if it is the IPvX address owner
      or if Accept_Mode is True.  Otherwise, MUST NOT accept these
      packets.
```
(note: VRRPv2 did not have an accept mode).

So according to the RFC, unless the priority of the VRRP instance is 255, packets sent to the virtual IP addresses should not be accepted (VRRP is designed for routeing purposes, and the only purpose in that case of the virtual IP address being configured is to allow ARP requests to be responded to), unless accept_mode is set.

keepalived implements this, but only if `strict_mode` or `no_accept` is configured. If you would like to post your configuration, we can check whether the keepalived configuration is stopping pings being received.

There could be other reasons for the pings not working, such as firewall settings, but that is beyond the scope of keepalived.

**Comment by yarcl** [comment]
This is my keepalived configuration below : 
`#ConfigurationFile for keepalived  
global_defs {  
notification_email {                        ######定义接受邮件的邮箱  
 yarcl@sina.cn
       }  
 notification_email_from localhost@aa.com    ######定义发送邮件的邮箱  
 smtp_server smtp.163.com  
 smtp_connect_timeout 10  
}  
vrrp_script check_nginx {                   ######定义监控nginx的脚本  
    script "/root/check_nginx.sh"  
    interval 2                            ######监控时间间隔  
    weight 2                              ######负载参数  
    }  
vrrp_instance vrrptest {                 ######定义vrrptest实例  
       state MASTER              ######服务器状态  
 interface eth0                     ######使用的接口  
       virtual_router_id 51               ######虚拟路由的标志，一组lvs的虚拟路由标识必须相同，这样才能切换  
       priority 150                       ######服务启动优先级，值越大，优先级越高，BACKUP 不能大于MASTER  
       advert_int 1                        ######服务器之间的存活检查时间  
authentication {  
       auth_type PASS                     ######认证类型  
       auth_pass 123456              ######认证密码，一组lvs 服务器的认证密码必须一致  
}  
track_script {                              ######执行监控nginx进程的脚本  
    check_nginx  
    }  
virtual_ipaddress {                         ######虚拟IP地址  
        10.25.95.245              ###飘移地址,行内空出来的ip
}  
}`
Is there any problems in this configuration file?I had tried many times, but is doesn't work.

Should I post the System configuration there?

**Comment by yarcl** [comment]
Ok， It was system configuration's problem and was fixed by my colleague， But I don't know how to fix it， he said it was dhcp's problem. Thanks @pqarmitage a lot.

**Comment by pqarmitage** [comment]
I believe your configuration is:
```
#Configuration File for keepalived

global_defs {
	notification_email
	{		 ######定义接受邮件的邮箱
		yarcl@sina.cn
	}

 	notification_email_from localhost@aa.com ######定义发送邮件的邮箱

 	smtp_server smtp.163.com
 	smtp_connect_timeout 10
}

vrrp_script check_nginx { ######定义监控nginx的脚本
	script "/root/check_nginx.sh"
	interval 2 ######监控时间间隔
	weight 2 ######负载参数
}

vrrp_instance vrrptest { ######定义vrrptest实例
	state MASTER ######服务器状态
	interface eth0 ######使用的接口
	virtual_router_id 51 ######虚拟路由的标志，一组lvs的虚拟路由标识必须相同，这样才能切换
	priority 150 ######服务启动优先级，值越大，优先级越高，BACKUP 不能大于MASTER
	advert_int 1 ######服务器之间的存活检查时间
	authentication {
		auth_type PASS ######认证类型
		auth_pass 123456 ######认证密码，一组lvs 服务器的认证密码必须一致
	}
 	track_script { ######执行监控nginx进程的脚本
 		check_nginx
 	}
	virtual_ipaddress { ######虚拟IP地址
		10.25.95.245 ###飘移地址,行内空出来的ip
	}
}
```
(the formatting seemed to somehow get lost when you posted it).

I have run the above configuration, and I can successfully ping 10.25.95.245 from another machine, so the problem does not appear to be you keepalived configuration.

I think the best way to diagnose the problem is to try it without keepalived, and make sure that works first. Execute the following: `ip addr add 10.25.95.245/32 dev eth0` and see if you can ping that address from a remote machine. If you can't ping it, then it indicates that the problem is not related to keepalived, but is likely to be a routing or firewalling issue, or some other network configuration.

**Comment by yarcl** [comment]
Yes, this problem was occured by network configuration, the firewall was not running. Problem has been solved by my colleague, Thank you @pqarmitage  very much.

**Comment by xiangwangcheng** [comment]
@yarcl  Would you please share the solution, I encountered the same problem.

**Comment by pqarmitage** [comment]
@xiangwangcheng The solution very much depends on your configuration, what version of keepalived you are using, and whether the iptables command creates nftables configuration or iptables configuration.

First of all you need to ensure that you don't have `strict` or `no_accept` configured for your VRRP instance. Then you need to ensure that keepalived isn't creating any nftables or iptables configuration to block packets to the virtual ip address. Finally, as described in https://github.com/acassen/keepalived/issues/1142#issuecomment-470011274 try the ping without keepalived running but after you have manually added the VIP.

**Comment by joshuaboniface** [comment]
>So according to the RFC, unless the priority of the VRRP instance is 255, packets sent to the virtual IP addresses should not be accepted (VRRP is designed for routeing purposes, and the only purpose in that case of the virtual IP address being configured is to allow ARP requests to be responded to), unless accept_mode is set.

Thank you @pqarmitage for mentioning this. I upgraded a number of keepalived instances with `vrrp_strict` enabled from Debian 10 Buster (2.0.10) to Debian 11 Bullseye (2.1.5) and ran into this issue; I had the `priority` values as simple multiples of 100, so both nodes would reject traffic. Not sure why it worked before, but afterwards it would not.

Posting this to help the next person who upgrades and runs into this, as this was my third or fourth Google result.

**Comment by pqarmitage** [comment]
@joshuaboniface The valid range of priorities is 1 to 255. If you set a priority of 256 or higher, you should see the following messages in your system log:
```
(xbl2.conf.conf: Line 29) number '256' outside range [1, 255]
(xbl2.conf: Line 36) (VIP_4) Priority not valid! must be between 1 & 255. Using default 100
```
so if you set a priority >= 256, that VRRP instance will run with a priority of 100, which explains why you are observing `priorities are now reversed, with lower priorities being preferred for the master state`.

I acknowledge that the keepalived.conf(5) man page does not state the valid range of priorities, and I will update it to do so. Nevertheless the VRRP RFCs (the current one is RFC5798) do clearly state this.

I will change the keepalived.conf(5) man page to state:
```
    # for electing MASTER, highest priority wins.
    # To be MASTER, it is recommended to make this 50 more than on
    # other machines. All systems should have different priorities
    # in order to make behaviour deterministic. If you want to stop
    # a higher priority instance taking over as master when it starts,
    # configure no_preempt rather than using equal priorities.
    # The valid range of values for priority is [1-255], with priority
    # 255 meaning "address owner". If no_accept is configured (or vrrp_strict
    # which also sets no_accept mode), then unless the vrrp_instance has
    # priority 255, the system will not receive packets addressed to the
    # VIPs/eVIPs, and the VIPs/eVIPs can only be used for routeing purposes.
    # Further, if an instance has priority 255 configured, the priority cannot
    # be reduced by track_scripts, track_process etc, and likewise
    # track_scripts etc cannot increase the priority to 255 if the configured
    # priority is not 255.
    \fBpriority \fR100
```

This should also explain the behaviour you are observing when you configure `vrrp_strict` which disables accept mode (accept mode is explained in keepalived.conf(5) man page, just search for `accept`). Disabling accept mode means that packets addressed to the VIPs/eVIPs will be dropped by the system (VRRP is designed for use by routers), but setting the priority to 255 means that accept mode cannot be disabled. So if accept mode is disabled and the priority is less that 255, the system will not accept packets addressed to the VIPs/eVIPs see [RFC 5798 para 6.1 Accept mode](https://datatracker.ietf.org/doc/html/rfc5798#section-6.1).

So far as I can see from your description above, everything is working as designed. You state that any priority other than 255 is broken, and that there is a bug in how `vrrp_strict` is working, but you haven't described what you are observing that brings you to that conclusion. If you could describe exactly what is not working, e.g.
  if `vrrp_strict` is set and the priority is not 255 then the system does not respond to pings being sent to a VIP when in master state
then we can address the issues further.
